File: ruby/binary-search-tree/binary_search_tree.rb
class Bst
  attr_reader :data, :left, :right

  def initialize(d)
    @data = d
    @left = nil
    @right = nil
  end

  def insert(d)
    if d <= data
      insert_into(:left, d)
    else
      insert_into(:right, d)
    end
  end

  def insert_into(dst, data)
    node = instance_variable_get("@#{dst}")
    if node
      node.insert(data)
    else
      instance_variable_set "@#{dst}", Bst.new(data)
    end
  end

  def each(&block)
    traverse_in_order(&block)
  end

  def traverse_in_order
    left.traverse_in_order { |node| yield node } if left
    yield data
    right.traverse_in_order { |node| yield node } if right
  end
end

File: ecmascript/binary-search-tree/binary-search-tree.js
class BinarySearchTree {
  constructor(data, left = null, right = null) {
    this.data = data;
    this.left = left;
    this.right = right;
  }

  insert(val) {
    let { data, left, right } = this;
    if (val <= data) {
      left
        ? left.insert(val)
        : this.left = new BinarySearchTree(val);
    } else {
      right
        ? right.insert(val)
        : this.right = new BinarySearchTree(val);
    }
  }

  each(func) {
    const { left, right, data } = this;
    if (left) left.each(func);
    func(data);
    if (right) right.each(func);
  }
}

export default BinarySearchTree;
